### ISSUE TITLE
Discuss providing "Installation Packages"

### ISSUE THREAD

**Issue by mikeschinkel** [wplib/wplib-box#118]
## Update:

See the [**strawman proposal below**](#issuecomment-239644309).
## Original:

Based on the request for PhpMyAdmin I am thinking about an idea we might call _"Installation Packages"_ or something similar.  

The idea would be for things we do not want to install on the machine itself we could allow for installation packages to be created and hosted somewhere like `installers.wplib.org`.  We could create them but we could also encourage other people to create them and make them available like packagist.org makes PHP packages available.

Basically they would be shell scripts that would add a feature to our box that people could install via our API, our CLI and/or our console.

Thoughts?

**Comment by mikeschinkel** [comment]
See 136#issuecomment-234421556 where I discussed some ideas about Installable Packages.

**Comment by mikeschinkel** [comment]
Potential Installable Packages:

- [MariaDB](https://mariadb.org/)
- [Percona Server for MySQL](https://www.percona.com/software/mysql-database/percona-server)
- [Apache](https://httpd.apache.org/)
- [Memcached](https://memcached.org/)
- [phpMyAdmin](https://www.phpmyadmin.net/)
- [HHVM](http://hhvm.com/)
- [ElasticSearch](https://www.elastic.co/products/elasticsearch)

*Please others propose ideas for this list.*

**Comment by mikeschinkel** [comment]
Having had time to consider this more here is a [**strawman proposal**](https://www.youtube.com/watch?v=JcN1i_9WV8c):

1. Add a `box install-package {package}` command to the [**Box CLI**](https://github.com/wplib/box-scripts/tree/master/guest/cli). _(Later the command might evolve to become `box package install {package}`.)_

1. An _"Installable Package"_ might be a tar file with a `.boxpkg` extension. 

1. A `.boxpkg` file could contain an `instructions.json` file that contain a list of tasks to perform such as `'box-command'`, `'download-plugin'` or `'install-dpkg'`, etc.

1. A `.boxpkg` could also contain a `/files/` folder that could contain the files that would be needed to be installed.

1. One common contained file type would likely be [`.deb` files](https://www.debian.org/doc/manuals/debian-faq/ch-pkg_basics.en.html) that [can be retrieved](http://unix.stackexchange.com/a/72778/144192) by using the `apt-get download` command. 

1. The person or team building the Installable Package would use `apt-get download` to retrieve the files needed for their `.boxpkg` since the idea is that the `apt-get` command would never actually be use during `box install-package`.

1. The `box install-package` command would likely often simply use `dpkg -i {.deb_file}` to install the package, assuming that the `.deb` file installs in a manner compatible with our directory layout.

1. We may even decide to pre-populate the apt-cache in `/var/cache/apt/archives/` with selected optional packages that we think many people will want to install.

1. We might actually decide to use `apt-get` and set up [our own apt-proxy](http://askubuntu.com/a/116596/486620) inside the box or on the web, and/or [use `--no-upgrade`](http://www.tecmint.com/useful-basic-commands-of-apt-get-and-apt-cache-for-package-management/) when running `apt-get` so an not to accidentally break a working box.

1. In the case where a `.deb` file does not install in a manner compatible with our directory layout we may just unpack the `.deb` file, modify as necessary, and then repackage as required into our package format.

1. Other Installable Packages could be as simple as a tarred `instructions.json` file that contains the task `'download-plugin'` with the URL where the plugin is found, and then `'install-plugin'` for that downloaded plugin.

1. Still other Installable Packages for which there is no available `.deb` file to be found could provide a `run-script` task that would do what is needed to install the `.boxpkg` file.

1. And &mdash; this just in &mdash; we could also possibly support [Snaps](https://techcrunch.com/2016/06/14/ubuntus-container-style-snap-app-packages-now-work-on-other-linux-distributions/) as a contained type of package that could be installed using Snap-native functionality.

1. We could create an Installable Package repository at [github.com/wplib/box-packages](https://github.com/wplib/box-packages) that could contain a `packages.json` file listing the official WPLib Box packages, and also any 3rd party developed packages.

1. In `packages.json` packages could be listed by a single segment slug such as 
`'php-7.1-hummingbird'` where _"hummingbird"_ could be [a major version](https://github.com/wplib/wplib-box/issues/155) of WPLib Box. 

1. However if a 3rd party such as [Mario Peshev](http://devwp.eu/about/) implemented an Installable Package for MongoDB it might be named `'devwp/mongodb-3.2-hummingbird'`.

**Comment by mikeschinkel** [comment]
I think the obvious question to counter the above strawman proposal is: 

> _"**Why not just use `apt-get`?**"_

This is why prefaced with the preference that _**installing a package should never break the box**_, if at all possible:

1. `apt-get` is a general purpose package manager meaning that it has to handle orders of magnitude more combinations of software than our box, and testing all those combinations is just not viable whereas it could be viable for us to test the most commonly used combinations for our box.

1. `apt-get` is a general purpose package manager meaning also that it has to handle orders of magnitude more combinations of software than our box, and testing all those combinations is just not viable whereas it could be viable for us to test the most commonly used combinations for our box.

1. `apt-get` is a general purpose package manager meaning also if we just use it we'd have no easy way to implement our own box-specific packages.

1. Box-specific packages can have benefit because it makes them much easier to create.  For example, a single `instructions.json` file that just specifies the two (2) tasks 'download-plugin'  and 'install-plugin' would be trivially easy for almost any box user to create.

1. We _could_ implement box-specific packages using `apt-get` _([the Debian Package manager](https://www.debian.org/doc/manuals/debian-faq/ch-pkg_basics.en.html))_  but implementing Debian packages is non-trivial and conflicts with the WPLib mission which is to provide the _Best Developer Experience_ possible.

1. We can easily **delegate** to any `apt-get` package so that we do not have to do the heavy lifting when we really do not need to.
 
1. Debian is not the only package manager now that [Snap App Packages](https://techcrunch.com/2016/06/14/ubuntus-container-style-snap-app-packages-now-work-on-other-linux-distributions/) are emerging so by having box-specific packages we could wrap and then delegate to more than one package manager.

1. `apt-get` has a duel concern of security and stability whereas the overriding concern for the box is stability since the box is designed only to run on your local machine. Thus `apt-get` repositories are updated frequently whereas the box package repository should only be updated after exhaustive stability testing.

**Comment by clubdeuce** [comment]
Why not just use docker?

**Comment by mikeschinkel** [comment]
@clubdeuce As I am currently envisioning it, an Installable Package **_may actually_** deploy a Docker container.  

But does it really make sense to require building a Docker container in order to install [a collection of plugins](https://github.com/wplib/box-scripts/issues/1), for example?   It seems to me that requiring a Docker container for that would be either overkill or not even applicable, but please check my logic on this.

**Comment by clubdeuce** [comment]
To further expound, I think this is the wrong road to take. By taking this approach, we are locking ourselves into needing a Debian based linux distribution.

**Comment by clubdeuce** [comment]
> But does it really make sense to require building a Docker container in order to install a collection of plugins, for example? It seems to me that requiring a Docker container for that would be either overkill or not even applicable, but please check my logic on this.

I was looking at the list of examples you provided at the top of this issue, all of which would be very simple to implement with Docker. A package of plugins would not work that way.

**Comment by mikeschinkel** [comment]
@clubdeuce 
> _we are locking ourselves into needing a Debian based linux distribution._

Hmm.  I see it as the opposite.  I am envisioning it Installable Packages would be release _(or "edition?")_ specific.  When we come out with a new release/edition existing packages would minimally need to be retested to run on the new release/edition if not completely reimplemented.

So while yes our [_"Hummingbird"_ release/edition](https://github.com/wplib/wplib-box/issues/155) may well be Debian, but that does not mean the _**"Peacock release/edition"**_ can't be CoreOS, for example. 

Also, think of our _"Installable Containers"_ as encapsulators just like methods can encapsulate access to private properties in an object.  So by having Installable Containers in front of `.deb` packages, Docker containers, `.zip` files, etc. we get to have logic that drives the installation.

**Comment by mikeschinkel** [comment]
@clubdeuce Of course this is all still very open for debate, so I am expecting you will aim to shoot any and all holes in the idea otherwise we may go down a wrong path.

**Comment by clubdeuce** [comment]
Closed by #219
